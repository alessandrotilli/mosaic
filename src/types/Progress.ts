import { IBase } from "./Base";

type ProgressType = "circular" | "linear";

export type ProgressProps = IBase & {
  type: ProgressType;
  withLabel?: boolean;
  color?: "primary" | "success" | "error";
  variant?: "determinate" | "indeterminate";
  value?: number;
};
