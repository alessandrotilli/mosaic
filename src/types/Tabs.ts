import { IBase } from "./Base";
import { ButtonColor } from "./Button";

export type TabType = {
  label: string;
  wrapped?: boolean;
  content?: React.ReactNode;
  disabled?: boolean;
  textColor?: ButtonColor;
};

export type TabsType = IBase & {
  tabList: TabType[];
  color?: ButtonColor;
  orientation?: "horizontal" | "vertical";
};
