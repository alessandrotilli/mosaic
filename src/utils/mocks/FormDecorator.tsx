import { useCallback, useEffect, useState } from "react";
import { DecoratorFn } from "@storybook/react";

export type FormValue = boolean | number | string | null;

const FormDecorator: DecoratorFn = (Story, { args: { onChange: externalOnChange, value: externalValue, ...args } }) => {
  const [value, setValue] = useState(externalValue);

  useEffect(() => {
    if (externalValue !== value) {
      setValue(externalValue);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [externalValue]);

  const onChange = useCallback(
    (value: FormValue) => {
      externalOnChange && externalOnChange(value);
      setValue(value);
    },
    [externalOnChange]
  );

  return Story({
    args: {
      ...args,
      onChange,
      value,
    },
  });
};

export default FormDecorator;
