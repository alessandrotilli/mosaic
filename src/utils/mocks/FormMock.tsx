import React, { cloneElement, FC, PropsWithChildren, ReactElement, useMemo, useState } from "react";
import { DecoratorFn } from "@storybook/react";

type IFormValue = boolean | number | string | null;

interface FormMockType {
  onChangePropName?: string;
  value: IFormValue;
  valuePropName?: string;
}

const FormMock: FC<PropsWithChildren<FormMockType>> = ({
  children,
  onChangePropName = "onChange",
  value: externalValue,
  valuePropName = "value",
}) => {
  const [value, setValue] = useState(externalValue);

  const wrappedFormElement = useMemo(
    () =>
      cloneElement(children as ReactElement<any>, {
        [onChangePropName]: setValue,
        [valuePropName]: value,
      }),
    [children, onChangePropName, value, valuePropName]
  );

  return wrappedFormElement;
};

const formDecorator: DecoratorFn = (Story, { args }) => <FormMock value={args.value}>{Story()}</FormMock>;

export { formDecorator, FormMock };
