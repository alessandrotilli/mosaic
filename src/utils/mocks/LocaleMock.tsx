import React from "react";
import { DecoratorFn } from "@storybook/react";

import { LocalizedContextProvider } from "../../contexts/Localized";

export type ILocaleMock = "en" | "it-IT";

export enum MessageMock {
  button = "intl.message.button",
  cancel = "intl.message.cancel",
  checkbox = "intl.message.checkbox",
  confirm = "intl.message.confirm",
  inputNumber = "intl.message.inputNumber",
  inputText = "intl.message.inputText",
  subtitle = "intl.message.subtitle",
  switch = "intl.message.switch",
  title = "intl.message.title",
  typography = "typography",
}

type IMessagesMock = {
  [key in ILocaleMock]: {
    [key in MessageMock]: string;
  };
};

export const MESSAGES_MOCK: IMessagesMock = {
  en: {
    [MessageMock.button]: "Button",
    [MessageMock.cancel]: "Cancel",
    [MessageMock.checkbox]: "Checkbox",
    [MessageMock.confirm]: "Confirm",
    [MessageMock.inputNumber]: "Input Number",
    [MessageMock.inputText]: "Input Text",
    [MessageMock.subtitle]: "Subtitle",
    [MessageMock.switch]: "Switch",
    [MessageMock.title]: "Title",
    [MessageMock.typography]: "Typography",
  },
  "it-IT": {
    [MessageMock.button]: "Bottone",
    [MessageMock.cancel]: "Annulla",
    [MessageMock.checkbox]: "Spunta",
    [MessageMock.confirm]: "Conferma",
    [MessageMock.inputNumber]: "Numero in input",
    [MessageMock.inputText]: "Testo in input",
    [MessageMock.subtitle]: "Sottotitolo",
    [MessageMock.switch]: "Interruttore",
    [MessageMock.title]: "Titolo",
    [MessageMock.typography]: "Tipografia",
  },
};

const getLocalizedMessage = (key: MessageMock, locale: ILocaleMock): string => MESSAGES_MOCK[locale][key];

const localeDecorator: DecoratorFn = (Story) => (
  <LocalizedContextProvider localize={(key) => getLocalizedMessage(key as MessageMock, "it-IT")}>
    {Story()}
  </LocalizedContextProvider>
);

export { getLocalizedMessage, localeDecorator };
