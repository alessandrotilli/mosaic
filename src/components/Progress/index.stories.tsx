import { Meta, StoryObj } from "@storybook/react";

import { localeDecorator } from "../../utils/mocks/LocaleMock";
import getDocsPage from "../../utils/stories";

import Progress, { DATA_CY_DEFAULT } from ".";

const COMPONENT_NAME = "Progress";
Progress.displayName = COMPONENT_NAME;

const meta = {
  title: "Feedback/Progress",
  component: Progress,
  decorators: [localeDecorator],
  parameters: {
    docs: {
      ...getDocsPage({
        basedOn: {
          label: "MUI Progress Component",
          url: "https://mui.com/material-ui/react-progress/",
        },
        component: COMPONENT_NAME,
        e2eTestInfo: {
          dataCyDefault: DATA_CY_DEFAULT,
        },
      }),
    },
  },
} satisfies Meta<typeof Progress>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Circular: Story = {
  args: {
    type: "circular",
  },
};

export const CircularColor: Story = {
  args: {
    ...Circular.args,
    color: "error",
  },
};

export const CircularDeterminate: Story = {
  args: {
    ...Circular.args,
    variant: "determinate",
    value: 50,
  },
};

export const CircularInvalidValue: Story = {
  args: {
    ...Circular.args,
    value: -999,
  },
};

export const CircularWithLabel: Story = {
  args: {
    ...CircularDeterminate.args,
    withLabel: true,
  },
};

export const Linear: Story = {
  args: {
    type: "linear",
  },
};

export const LinearColor: Story = {
  args: {
    ...Linear.args,
    color: "error",
  },
};

export const LinearDeterminate: Story = {
  args: {
    ...Linear.args,
    variant: "determinate",
    value: 50,
  },
};

export const LinearInvalidValue: Story = {
  args: {
    ...Linear.args,
    value: 999,
  },
};

export const LinearWithLabel: Story = {
  args: {
    ...LinearDeterminate.args,
    withLabel: true,
  },
};
