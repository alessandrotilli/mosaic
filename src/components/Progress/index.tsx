import React, { FC, useMemo } from "react";
import { Box, LinearProgress as MUILinearProgress } from "@mui/material";
import MUICircularProgress from "@mui/material/CircularProgress";
import Typography from "@mui/material/Typography";

import { ProgressType } from "../../types/Progress";

export const DATA_CY_DEFAULT = "progress";

const Progress: FC<ProgressType> = ({
  dataCy = DATA_CY_DEFAULT,
  color = "primary",
  type = "Circular",
  withLabel,
  variant,
  value = 100,
}) => {
  const percent = useMemo(() => {
    if (value < 0) {
      return 0;
    } else if (value > 100) {
      return 100;
    }
    return value;
  }, [value]);

  if (type === "Circular") {
    return (
      <Box sx={{ position: "relative", display: "inline-flex" }}>
        <MUICircularProgress data-cy={dataCy} color={color} variant={variant} value={percent} />
        {withLabel && (
          <Box
            sx={{
              top: 0,
              left: 0,
              bottom: 0,
              right: 0,
              position: "absolute",
              display: "flex",
              alignItems: "center",
              justifyContent: "center",
            }}
          >
            <Typography variant="caption" component="div" color="text.secondary">
              {percent + `%`}
            </Typography>
          </Box>
        )}
      </Box>
    );
  }
  return (
    <Box sx={{ display: "flex", alignItems: "center" }}>
      <Box sx={{ width: "100%", mr: 1 }}>
        <MUILinearProgress data-cy={dataCy} color={color} variant={variant} value={percent} />
      </Box>
      {withLabel && (
        <Box sx={{ minWidth: 35 }}>
          <Typography variant="body2" color="text.secondary">
            {percent + `%`}
          </Typography>
        </Box>
      )}
    </Box>
  );
};

export default Progress;
