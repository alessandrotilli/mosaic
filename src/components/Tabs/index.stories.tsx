import React from "react";
import { ComponentMeta, ComponentStory } from "@storybook/react";

import { localeDecorator } from "../../utils/mocks/LocaleMock";
import getDocsPage from "../../utils/stories";
import Progress from "../Progress";

import Tabs, { DATA_CY_DEFAULT } from ".";

const COMPONENT_NAME = "Tabs";
Tabs.displayName = COMPONENT_NAME;

export default {
  title: "Navigation/Tabs",
  component: Tabs,
  decorators: [localeDecorator],
  parameters: {
    docs: {
      ...getDocsPage({
        basedOn: {
          label: "MUI Tabs Component",
          url: "https://mui.com/material-ui/react-progress/",
        },
        component: COMPONENT_NAME,
        e2eTestInfo: {
          dataCyDefault: DATA_CY_DEFAULT,
        },
      }),
    },
  },
} as ComponentMeta<typeof Tabs>;

const Template: ComponentStory<typeof Tabs> = (args) => <Tabs {...args} dataCy={DATA_CY_DEFAULT} />;

export const Primary = Template.bind({});
Primary.args = {
  labelList: [{ label: "PAGE1" }, { label: "PAGE2" }, { label: "PAGE3" }],
};

export const Color = Template.bind({});
Color.args = { ...Primary.args, color: "secondary" };

export const Custom = Template.bind({});
Custom.args = {
  labelList: [
    { label: "PAGE1" },
    { label: "PAGE2" },
    { label: "PAGE3", children: <Progress type="Circular" withLabel={false} /> },
  ],
};

export const Orientation = Template.bind({});
Orientation.args = { ...Custom.args, orientation: "vertical" };
