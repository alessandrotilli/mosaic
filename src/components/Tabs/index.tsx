import React, { FC } from "react";
import Box from "@mui/material/Box";
import MUITab from "@mui/material/Tab";
import MUITabs from "@mui/material/Tabs";

import { TabsType } from "../../types/Tabs";
import Typography from "../Typography";

export const DATA_CY_DEFAULT = "tabs";

const Tabs: FC<TabsType> = ({ dataCy = DATA_CY_DEFAULT, labelList, color = "primary" }) => {
  const [value, setValue] = React.useState(0);

  const handleChange = (event: React.SyntheticEvent, newValue: number) => {
    setValue(newValue);
  };
  return (
    <Box>
      <MUITabs data-cy={dataCy} value={value} onChange={handleChange} textColor={color} indicatorColor={color}>
        {labelList.map((i, index) => {
          return <MUITab label={i.label} wrapped={i.wrapped} key={index} />;
        })}
      </MUITabs>
      {labelList[value].children && (
        <div role="tabpanel" id={`simple-tabpanel-${value}`}>
          <Box sx={{ p: 3 }}>
            <Typography>{labelList[value].children}</Typography>
          </Box>
        </div>
      )}
    </Box>
  );
};

export default Tabs;
