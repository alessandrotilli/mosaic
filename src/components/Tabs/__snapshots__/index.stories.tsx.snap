// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navigation/Tabs Color smoke-test 1`] = `
<div class="MuiBox-root css-0">
  <div class="MuiTabs-root css-1ujnqem-MuiTabs-root"
       data-cy="tabs"
  >
    <div class="MuiTabs-scrollableX MuiTabs-hideScrollbar css-oqr85h"
         style="width: 99px; height: 99px; position: absolute; top: -9999px; overflow: scroll;"
    >
    </div>
    <div class="MuiTabs-scroller MuiTabs-hideScrollbar MuiTabs-scrollableX css-69z67c-MuiTabs-scroller"
         style="margin-bottom: 0px;"
    >
      <div class="MuiTabs-flexContainer css-heg063-MuiTabs-flexContainer"
           role="tablist"
      >
        <button class="MuiButtonBase-root MuiTab-root MuiTab-textColorSecondary Mui-selected css-1a4cg4j-MuiButtonBase-root-MuiTab-root"
                tabindex="0"
                type="button"
                role="tab"
                aria-selected="true"
        >
          PAGE1
          <span class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root">
          </span>
        </button>
        <button class="MuiButtonBase-root MuiTab-root MuiTab-textColorSecondary css-1a4cg4j-MuiButtonBase-root-MuiTab-root"
                tabindex="-1"
                type="button"
                role="tab"
                aria-selected="false"
        >
          PAGE2
          <span class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root">
          </span>
        </button>
        <button class="MuiButtonBase-root MuiTab-root MuiTab-textColorSecondary css-1a4cg4j-MuiButtonBase-root-MuiTab-root"
                tabindex="-1"
                type="button"
                role="tab"
                aria-selected="false"
        >
          PAGE3
          <span class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root">
          </span>
        </button>
      </div>
      <span class="MuiTabs-indicator css-11yukd5-MuiTabs-indicator"
            style="left: 0px; width: 90px;"
      >
      </span>
    </div>
  </div>
  <div role="tabpanel"
       id="simple-tabpanel-0"
       aria-labelledby="horizontal-tab-0"
       style="width: 100%;"
  >
    PAGE1
  </div>
</div>
`;

exports[`Navigation/Tabs Orientation smoke-test 1`] = `
<div class="MuiBox-root css-k008qs">
  <div class="MuiTabs-root MuiTabs-vertical css-1g4ac1o-MuiTabs-root"
       data-cy="tabs"
  >
    <div class="MuiTabs-scroller MuiTabs-fixed css-jpln7h-MuiTabs-scroller"
         style="overflow: hidden; margin-right: 0px;"
    >
      <div aria-orientation="vertical"
           class="MuiTabs-flexContainer MuiTabs-flexContainerVertical css-lfwcke-MuiTabs-flexContainer"
           role="tablist"
      >
        <button class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary Mui-selected css-1h9z7r5-MuiButtonBase-root-MuiTab-root"
                tabindex="0"
                type="button"
                role="tab"
                aria-selected="true"
        >
          PAGE1
          <span class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root">
          </span>
        </button>
        <button class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary css-1h9z7r5-MuiButtonBase-root-MuiTab-root"
                tabindex="-1"
                type="button"
                role="tab"
                aria-selected="false"
        >
          PAGE2
          <span class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root">
          </span>
        </button>
        <button class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary css-1h9z7r5-MuiButtonBase-root-MuiTab-root"
                tabindex="-1"
                type="button"
                role="tab"
                aria-selected="false"
        >
          PAGE3
          <span class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root">
          </span>
        </button>
      </div>
      <span class="MuiTabs-indicator css-10d9dml-MuiTabs-indicator"
            style="top: 0px; height: 48px;"
      >
      </span>
    </div>
  </div>
  <div role="tabpanel"
       id="simple-tabpanel-0"
       aria-labelledby="vertical-tab-0"
       style="width: 100%;"
  >
    PAGE1
  </div>
</div>
`;

exports[`Navigation/Tabs Primary smoke-test 1`] = `
<div class="MuiBox-root css-0">
  <div class="MuiTabs-root css-1ujnqem-MuiTabs-root"
       data-cy="tabs"
  >
    <div class="MuiTabs-scrollableX MuiTabs-hideScrollbar css-oqr85h"
         style="width: 99px; height: 99px; position: absolute; top: -9999px; overflow: scroll;"
    >
    </div>
    <div class="MuiTabs-scroller MuiTabs-hideScrollbar MuiTabs-scrollableX css-69z67c-MuiTabs-scroller"
         style="margin-bottom: 0px;"
    >
      <div class="MuiTabs-flexContainer css-heg063-MuiTabs-flexContainer"
           role="tablist"
      >
        <button class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary Mui-selected css-1h9z7r5-MuiButtonBase-root-MuiTab-root"
                tabindex="0"
                type="button"
                role="tab"
                aria-selected="true"
        >
          PAGE1
          <span class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root">
          </span>
        </button>
        <button class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary css-1h9z7r5-MuiButtonBase-root-MuiTab-root"
                tabindex="-1"
                type="button"
                role="tab"
                aria-selected="false"
        >
          PAGE2
          <span class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root">
          </span>
        </button>
        <button class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary css-1h9z7r5-MuiButtonBase-root-MuiTab-root"
                tabindex="-1"
                type="button"
                role="tab"
                aria-selected="false"
        >
          PAGE3
          <span class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root">
          </span>
        </button>
      </div>
      <span class="MuiTabs-indicator css-1aquho2-MuiTabs-indicator"
            style="left: 0px; width: 90px;"
      >
      </span>
    </div>
  </div>
  <div role="tabpanel"
       id="simple-tabpanel-0"
       aria-labelledby="horizontal-tab-0"
       style="width: 100%;"
  >
    PAGE1
  </div>
</div>
`;
